File: tests/xcalablemp/global-view/gmove/F/gmove_inout_3.f90
module mod0

  integer, parameter :: N = 4

!$xmp nodes p0(8)

!$xmp nodes p1(2,2) = p0(1:4)
!$xmp nodes p2(2,2) = p0(5:8)

!$xmp template t1(N,N,N)
!$xmp distribute t1(*,block,block) onto p1

  integer a(N,N,N)
!$xmp align a(i,j,k) with t1(i,j,k)
!$xmp shadow a(0,2:1,1:0)

!$xmp template t2(N,N,N)
!$xmp distribute t2(block,cyclic,*) onto p2

  integer b(N,N,N)
!$xmp align b(i,j,k) with t2(i,j,k)
!$xmp shadow b(0:1,0,0)

end module mod0


!--------------------------------------------------------
program test

  use mod0

#ifdef _MPI3
  call gmove_in
  call gmove_in_async
  call gmove_out
  call gmove_out_async

!$xmp task on p0(1)
  write(*,*) "PASS"
!$xmp end task
#else
  write(*,*) "Skipped"
#endif

end program test


!--------------------------------------------------------
subroutine gmove_in

integer :: result = 0

  use mod0

!$xmp loop (i,j,k) on t1(i,j,k)
  do k = 1, N
     do j = 1, N
        do i = 1, N
           a(i,j,k) = 777
        end do
     end do
  end do

!$xmp loop (i,j,k) on t2(i,j,k)
  do k = 1, N
     do j = 1, N
        do i = 1, N
           b(i,j,k) = i*10000 + j *100 + k
        end do
     end do
  end do

!$xmp barrier

!$xmp task on p1

#ifdef _MPI3
!$xmp gmove in
  a = b
#endif

!$xmp end task

!$xmp barrier

!$xmp loop (i,j,k) on t1(i,j,k) reduction(+:result)
  do k = 1, N
     do j = 1, N
        do i = 1, N
           if (a(i,j,k) /= i*10000 + j*100 + k) then
              write(*,*) "(", xmp_node_num(), ")", i, j, k, a(i,j,k)
              result = 1
           end if
        end do
     end do
  end do

!$xmp task on p0(1) nocomm
  if (result /= 0) then
     write(*,*) "ERROR in gmove_in"
     call exit(1)
  endif
!$xmp end task

end subroutine gmove_in


!--------------------------------------------------------
subroutine gmove_in_async

integer :: result = 0

  use mod0

!$xmp loop (i,j,k) on t1(i,j,k)
  do k = 1, N
     do j = 1, N
        do i = 1, N
           a(i,j,k) = 777
        end do
     end do
  end do

!$xmp loop (i,j,k) on t2(i,j,k)
  do k = 1, N
     do j = 1, N
        do i = 1, N
           b(i,j,k) = i*10000 + j *100 + k
        end do
     end do
  end do

!$xmp barrier

!$xmp task on p1

#ifdef _MPI3
!$xmp gmove in async(10)
  a = b
#endif

!$xmp wait_async(10)

!$xmp end task

!$xmp barrier

!$xmp loop (i,j,k) on t1(i,j,k) reduction(+:result)
  do k = 1, N
     do j = 1, N
        do i = 1, N
           if (a(i,j,k) /= i*10000 + j*100 + k) then
              !write(*,*) "(", xmp_node_num(), ")", i, j, k, a(i,j,k)
              result = 1
           end if
        end do
     end do
  end do

!$xmp task on p0(1) nocomm
  if (result /= 0) then
     write(*,*) "ERROR in gmove_in"
     call exit(1)
  endif
!$xmp end task

end subroutine gmove_in_async


!--------------------------------------------------------
subroutine gmove_out

integer :: result = 0

  use mod0

!$xmp loop (i,j,k) on t1(i,j,k)
  do k = 1, N
     do j = 1, N
        do i = 1, N
           a(i,j,k) = 777
        end do
     end do
  end do

!$xmp loop (i,j,k) on t2(i,j,k)
  do k = 1, N
     do j = 1, N
        do i = 1, N
           b(i,j,k) = i*10000 + j *100 + k
        end do
     end do
  end do

!$xmp barrier

!$xmp task on p2

#ifdef _MPI3
!$xmp gmove out
  a = b
#endif

!$xmp end task

!$xmp barrier

!$xmp loop (i,j,k) on t1(i,j,k) reduction(+:result)
  do k = 1, N
     do j = 1, N
        do i = 1, N
           if (a(i,j,k) /= i*10000 + j*100 + k) then
              !write(*,*) "(", xmp_node_num(), ")", i, j, k, a(i,j,k)
              result = 1
           end if
        end do
     end do
  end do

!$xmp task on p0(1)
  if (result /= 0) then
     write(*,*) "ERROR in gmove_in"
     call exit(1)
  endif
!$xmp end task

end subroutine gmove_out


!--------------------------------------------------------
subroutine gmove_out_async

integer :: result = 0

  use mod0

!$xmp loop (i,j,k) on t1(i,j,k)
  do k = 1, N
     do j = 1, N
        do i = 1, N
           a(i,j,k) = 777
        end do
     end do
  end do

!$xmp loop (i,j,k) on t2(i,j,k)
  do k = 1, N
     do j = 1, N
        do i = 1, N
           b(i,j,k) = i*10000 + j *100 + k
        end do
     end do
  end do

!$xmp barrier

!$xmp task on p2

#ifdef _MPI3
!$xmp gmove out async(10)
  a = b
#endif

!$xmp wait_async(10)

!$xmp end task

!$xmp barrier

!$xmp loop (i,j,k) on t1(i,j,k) reduction(+:result)
  do k = 1, N
     do j = 1, N
        do i = 1, N
           if (a(i,j,k) /= i*10000 + j*100 + k) then
              !write(*,*) "(", xmp_node_num(), ")", i, j, k, a(i,j,k)
              result = 1
           end if
        end do
     end do
  end do

!$xmp task on p0(1)
  if (result /= 0) then
     write(*,*) "ERROR in gmove_in"
     call exit(1)
  endif
!$xmp end task

end subroutine gmove_in_async

File: tests/XMP/global-view/gmove/C/gmove_inout_3.c
#include <stdio.h>
#include <stdlib.h>
#include "xmp.h"

#define N 4

#pragma xmp nodes p0[8]

#pragma xmp nodes p1[2][2] = p0[0:4]
#pragma xmp nodes p2[2][2] = p0[4:4]

#pragma xmp template t1[N][N][N]
#pragma xmp distribute t1[block][block][*] onto p1

int a[N][N][N];
#pragma xmp align a[i][j][k] with t1[k][j][i]
#pragma xmp shadow a[0][2:1][1:0]

#pragma xmp template t2[N][N][N]
#pragma xmp distribute t2[*][cyclic][block] onto p2

int b[N][N][N];
#pragma xmp align b[i][j][k] with t2[k][j][i]
#pragma shadow b[0:1][0][0]


//--------------------------------------------------------
void gmove_in(){

  int result = 0;

#pragma xmp loop on t1[k][j][i]
  for (int i = 0; i < N; i++){
    for (int j = 0; j < N; j++){
      for (int k = 0; k < N; k++){
	a[i][j][k] = 777;
      }
    }
  }

#pragma xmp loop on t2[k][j][i]
  for (int i = 0; i < N; i++){
    for (int j = 0; j < N; j++){
      for (int k = 0; k < N; k++){
	b[i][j][k] = i*10000 + j *100 + k;
      }
    }
  }

#pragma xmp barrier

#pragma xmp task on p1
  {
#ifdef _MPI3
#pragma xmp gmove in
    a[:][:][:] = b[:][:][:];
#endif
  }

#pragma xmp barrier

#pragma xmp loop on t1[k][j][i] reduction(+:result)
  for (int i = 0; i < N; i++){
    for (int j = 0; j < N; j++){
      for (int k = 0; k < N; k++){
	if (a[i][j][k] != i*10000 + j*100 + k){
	  //printf("(%d) %d %d %d %d\n", xmp_node_num(), i, j, k, a[i][j][k]);
	  result = 1;
	}
      }
    }
  }

#pragma xmp task on p0[0]
  {
    if (result != 0){
      printf("ERROR in gmove_in\n");
      exit(1);
    }
  }

}


//--------------------------------------------------------
void gmove_in_async(){

  int result = 0;

#pragma xmp loop on t1[k][j][i]
  for (int i = 0; i < N; i++){
    for (int j = 0; j < N; j++){
      for (int k = 0; k < N; k++){
	a[i][j][k] = 777;
      }
    }
  }

#pragma xmp loop on t2[k][j][i]
  for (int i = 0; i < N; i++){
    for (int j = 0; j < N; j++){
      for (int k = 0; k < N; k++){
	b[i][j][k] = i*10000 + j *100 + k;
      }
    }
  }

#pragma xmp barrier

#pragma xmp task on p1
  {
#ifdef _MPI3
#pragma xmp gmove in async(10)
    a[:][:][:] = b[:][:][:];
#endif

#pragma xmp wait_async (10)
  }

#pragma xmp barrier

#pragma xmp loop on t1[k][j][i] reduction(+:result)
  for (int i = 0; i < N; i++){
    for (int j = 0; j < N; j++){
      for (int k = 0; k < N; k++){
	if (a[i][j][k] != i*10000 + j*100 + k){
	  //printf("(%d) %d %d %d %d\n", xmp_node_num(), i, j, k, a[i][j][k]);
	  result = 1;
	}
      }
    }
  }

#pragma xmp task on p0[0]
  {
    if (result != 0){
      printf("ERROR in gmove_in_async\n");
      exit(1);
    }
  }

}


//--------------------------------------------------------
void gmove_out(){

  int result = 0;

#pragma xmp loop on t1[k][j][i]
  for (int i = 0; i < N; i++){
    for (int j = 0; j < N; j++){
      for (int k = 0; k < N; k++){
	a[i][j][k] = 777;
      }
    }
  }

#pragma xmp loop on t2[k][j][i]
  for (int i = 0; i < N; i++){
    for (int j = 0; j < N; j++){
      for (int k = 0; k < N; k++){
	b[i][j][k] = i*10000 + j *100 + k;
      }
    }
  }

#pragma xmp barrier

#pragma xmp task on p2
  {
#ifdef _MPI3
#pragma xmp gmove out
    a[:][:][:] = b[:][:][:];
#endif
  }

#pragma xmp barrier

#pragma xmp loop on t1[k][j][i] reduction(+:result)
  for (int i = 0; i < N; i++){
    for (int j = 0; j < N; j++){
      for (int k = 0; k < N; k++){
	if (a[i][j][k] != i*10000 + j*100 + k){
	  //printf("(%d) %d %d %d %d\n", xmp_node_num(), i, j, k, a[i][j][k]);
	  result = 1;
	}
      }
    }
  }

#pragma xmp task on p0[0]
  {
    if (result != 0){
      printf("ERROR in gmove_out\n");
      exit(1);
    }
  }

}


//--------------------------------------------------------
void gmove_out_async(){

  int result = 0;

#pragma xmp loop on t1[k][j][i]
  for (int i = 0; i < N; i++){
    for (int j = 0; j < N; j++){
      for (int k = 0; k < N; k++){
	a[i][j][k] = 777;
      }
    }
  }

#pragma xmp loop on t2[k][j][i]
  for (int i = 0; i < N; i++){
    for (int j = 0; j < N; j++){
      for (int k = 0; k < N; k++){
	b[i][j][k] = i*10000 + j *100 + k;
      }
    }
  }

#pragma xmp barrier

#pragma xmp task on p2
  {
#ifdef _MPI3
#pragma xmp gmove out async(10)
    a[:][:][:] = b[:][:][:];
#endif

#pragma xmp wait_async (10)
  }

#pragma xmp barrier

#pragma xmp loop on t1[k][j][i] reduction(+:result)
  for (int i = 0; i < N; i++){
    for (int j = 0; j < N; j++){
      for (int k = 0; k < N; k++){
	if (a[i][j][k] != i*10000 + j*100 + k){
	  //printf("(%d) %d %d %d %d\n", xmp_node_num(), i, j, k, a[i][j][k]);
	  result = 1;
	}
      }
    }
  }

#pragma xmp task on p0[0]
  {
    if (result != 0){
      printf("ERROR in gmove_out_async\n");
      exit(1);
    }
  }

}


//--------------------------------------------------------
int main(){

#ifdef _MPI3
  gmove_in();
  gmove_in_async();
  gmove_out();
  gmove_out_async();

#pragma xmp task on p0[0]
  {
    printf("PASS\n");
  }
#else
  printf("Skipped\n");
#endif

  return 0;

}
